apps/dash-datashader/app.py

```diff
@@ -0,0 +1,316 @@
+    ):
+        x0 = selection["xaxis.range[0]"]
+        x1 = selection["xaxis.range[1]"]
+        sub_df = df[(df.Time >= x0) & (df.Time <= x1)]
+        num_pts = len(sub_df)
+        if num_pts < max_points:
+            number = "{:,}".format(
+                abs(int(selection["xaxis.range[1]"]) - int(selection["xaxis.range[0]"]))
+            )
+            number_print = " points selected between {0:,.4} and {1:,.4}".format(
+                selection["xaxis.range[0]"], selection["xaxis.range[1]"]
+            )
+        else:
+            number = "{:,}".format(
+                abs(int(selection["xaxis.range[1]"]) - int(selection["xaxis.range[0]"]))
+            )
+            number_print = " points selected. Select less than {0:}k \
+            points to invoke high-res scattergl trace".format(
+                max_points / 1000
+            )
+    else:
+        number = "0"
+        number_print = " points selected"
+    return [number, number_print]
+
+
+@app.callback(Output("graph-2", "figure"), [Input("graph-1", "relayoutData")])
+def selectionHighlight(selection):
+    new_fig2 = fig2.copy()
+    if (
+        selection is not None
+        and "xaxis.range[0]" in selection
+        and "xaxis.range[1]" in selection
+    ):
+        x0 = selection["xaxis.range[0]"]
+        x1 = selection["xaxis.range[1]"]
+        sub_df = df[(df.Time >= x0) & (df.Time <= x1)]
+        num_pts = len(sub_df)
+        if num_pts < max_points:
+            shape = dict(
+                type="rect",
+                xref="x",
+                yref="paper",
+                y0=0,
+                y1=1,
+                x0=x0,
+                x1=x1,
+                line={"width": 0},
+                fillcolor="rgba(165, 131, 226, 0.10)",
+            )
+
+            new_fig2["layout"]["shapes"] = [shape]
+        else:
+            new_fig2["layout"]["shapes"] = []
+    else:
+        new_fig2["layout"]["shapes"] = []
+    return new_fig2
+
+
+@app.callback(Output("graph-1", "figure"), [Input("graph-1", "relayoutData")])
+def draw_undecimated_data(selection):
+    new_fig1 = fig1.copy()
+    if (
+        selection is not None
+        and "xaxis.range[0]" in selection
+        and "xaxis.range[1]" in selection
+    ):
+        x0 = selection["xaxis.range[0]"]
+        x1 = selection["xaxis.range[1]"]
+        sub_df = df[(df.Time >= x0) & (df.Time <= x1)]
+        num_pts = len(sub_df)
+        if num_pts < max_points:
+            high_res_data = [
+                dict(
+                    x=sub_df["Time"],
+                    y=sub_df["Signal"],
+                    type="scattergl",
+                    marker=dict(sizemin=1, sizemax=30, color="#a3a7b0"),
+                )
+            ]
+            high_res_layout = new_fig1["layout"]
+            high_res = dict(data=high_res_data, layout=high_res_layout)
+        else:
+            high_res = fig1.copy()
+    else:
+        high_res = fig1.copy()
+    return high_res
```

Comment:
Small suggestion :)

I would try to reduce the amount of unnecessary `if-else`. 
I would do something like this. I would replace line `306-312`.
```python
            new_fig1 = fig1.copy()
            high_res_layout = new_fig1["layout"]
            return dict(data=high_res_data, layout=high_res_layout)
    return fig1.copy()
```

Reply:
@YunkeXiao I can't tell if you've addressed this ^ yet? I know it's not your code, but I think it's a good idea to avoid all of these repetitive `fig1.copy()` cases like @chriskfwoo suggested.